app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollView.kt

```diff
@@ -0,0 +1,266 @@
+    private var xStart = 0
+    private var roomsCount = NOT_INITIALIZED
+    private var roomNames: HorizontalScrollView? = null
+    private var displayColumnCount: Int
+    private var columnWidth: Int
+
+    // The currently displayed column index
+    var activeColumnIndex = 0
+        private set
+
+    init {
+        displayColumnCount = resources.getInteger(R.integer.max_cols)
+        columnWidth = 0
+        gestureDetector = GestureDetector(this.context, YScrollDetector())
+        setOnTouchListener(OnTouchListener())
+    }
+
+    fun getColumnWidth(): Int {
+        check(roomsCount != NOT_INITIALIZED) { "The \"roomsCount\" field must be initialized before invoking \"getColumnWidth\"." }
+        return columnWidth
+    }
+
+    inner class YScrollDetector : SimpleOnGestureListener() {
+
+        override fun onDown(e: MotionEvent): Boolean {
+            xStart = e.x.toInt()
+            activeColumnIndex = ((scrollX * displayColumnCount).toFloat() / measuredWidth).roundToInt()
+            return super.onDown(e)
+        }
+
+        override fun onFling(
+            e1: MotionEvent,
+            e2: MotionEvent,
+            velocityX: Float,
+            velocityY: Float
+        ): Boolean {
+            val normalizedAbsoluteVelocityX = abs(velocityX / resources.displayMetrics.density)
+            val columns = ceil((normalizedAbsoluteVelocityX / SWIPE_THRESHOLD_VELOCITY * 3).toDouble()).toInt()
+            val exceedsVelocityThreshold = normalizedAbsoluteVelocityX > SWIPE_THRESHOLD_VELOCITY
+
+            try {
+                //right to left
+                if (e1.x - e2.x > SWIPE_MIN_DISTANCE && exceedsVelocityThreshold) {
+                    scrollToColumn(activeColumnIndex + columns, false)
+                    return true
+                }
+                //left to right
+                else if (e2.x - e1.x > SWIPE_MIN_DISTANCE && exceedsVelocityThreshold) {
+                    scrollToColumn(activeColumnIndex - columns, false)
+                    return true
+                }
+            } catch (e: Exception) {
+                MyApp.LogDebug(LOG_TAG, "There was an error processing the Fling event:" + e.message)
+            }
+
+            return super.onFling(e1, e2, velocityX, velocityY)
+        }
+    }
+
+    override fun onInterceptTouchEvent(ev: MotionEvent): Boolean {
+        gestureDetector.onTouchEvent(ev)
+        return super.onInterceptTouchEvent(ev)
+    }
+
+    /**
+     * Sets the rooms count so it can be used to calculate
+     * the room column dimensions in the next layout phase.
+     */
+    fun setRoomsCount(@IntRange(from = 1) roomsCount: Int) {
+        this.roomsCount = roomsCount
+    }
+
+    fun scrollToColumn(col: Int, fast: Boolean) {
+        val max = if (this.columnWidth == 0) 0 else (getChildAt(0).measuredWidth - columnWidth) / columnWidth
+        val col = min(max(col, 0), max)
+
+        val scrollTo = col * columnWidth
```

Comment:
I would never shadow another variable on purpose. This could lead to unforseen error when refactoring. Why not `columnIndex`?

Reply:
Because the new Variable is just the old one but Clamped to a specific Range (between 0 and the calculated maximum Value). In Java, you could just change the value of the Parameter, but for some reason, Parameters are read-only in Kotlin. I'll change it to `columnIndex`

Reply:
In Java one would make parameters `final` to guarantee that the function does not modify the input which can be very dangerous. Thanks for the change.